dspy/streaming/streaming_listener.py

```diff
@@ -196,30 +203,98 @@ def receive(self, chunk: ModelResponseStream):
 
         if self.stream_start and chunk_message:
             # The stream is started, we keep returning the token until we see the start of the next field.
-            token = None
             self.field_end_queue.put(chunk_message)
 
+            token = None
             concat_message = "".join(self.field_end_queue.queue).strip()
-            if re.search(end_identifier, concat_message):
-                # The next field is identified, we can end the stream and flush out all tokens in the buffer.
-                self.stream_end = True
-                token = self.flush()
-                token = token.rstrip()  # Remove the trailing \n\n
-            elif not self._could_form_end_identifier(concat_message, adapter_name):
+
+            if not self._could_form_end_identifier(concat_message, adapter_name):
                 # Buffer cannot form end identifier, safe to flush out the tokens in the buffer.
                 token = self.flush()
             elif self.field_end_queue.qsize() > 10:
-                # Buffer could form end identifier, but we've exceeded max buffer size
-                # Yield the oldest token to prevent unbounded buffering
+                # We keep the last 10 tokens in the buffer if they can potentially form the end_identifier to avoid
+                # sending the DSPy bolilerplate tokens to users. 10 is a heuristic number that is sufficient to capture
+                # the end_identifier for all LMs.
                 token = self.field_end_queue.get()
 
-            if token:
+            if isinstance(settings.adapter, JSONAdapter):
+                # JSONAdapter uses partial json parsing to detect the end of the field we are listening to, instead of
+                # relying on the end_identifier.
+                return self._json_adapter_handle_stream_chunk(token, chunk_message)
+            else:
+                # Other adapters rely on the end_identifier to detect the end of the field we are listening to.
+                return self._default_handle_stream_chunk(token, end_identifier)
+
+    def _json_adapter_handle_stream_chunk(self, token: str, chunk_message: str) -> StreamResponse | None:
+        self.json_adapter_state["field_accumulated_tokens"] += chunk_message
```

Comment:
Why do we accumulate `chunk_message` instead of `token`?

Reply:
Basically token is the one we return, `chunk_message ` is the new message we are receiving.